src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -0,0 +1,303 @@
+/*-------------------------------------------------------------------------
+ *
+ * distributed_deadlock_detection.c
+ *
+ *  Functions for performing distributed deadlock detection.
+ *
+ * Copyright (c) 2017, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "funcapi.h"
+#include "miscadmin.h"
+
+#include "access/hash.h"
+#include "distributed/backend_data.h"
+#include "distributed/hash_helpers.h"
+#include "distributed/listutils.h"
+#include "distributed/lock_graph.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/transaction_identifier.h"
+#include "nodes/pg_list.h"
+#include "utils/hsearch.h"
+#include "utils/timestamp.h"
+
+
+typedef struct TransactionNode
+{
+	DistributedTransactionId transactionId;
+
+	/* list of TransactionNode that this distributed transaction is waiting for */
+	List *waitsFor;
+} TransactionNode;
+
+
+static char * WaitsForToString(List *waitsFor);
+static HTAB * FlattenWaitGraph(WaitGraph *waitGraph);
+static TransactionNode * GetTransactionNode(HTAB *distributedTransactionHash,
+											DistributedTransactionId *transactionId);
+static uint32 DistributedTransactionIdHash(const void *key, Size keysize);
+static int DistributedTransactionIdCompare(const void *a, const void *b, Size keysize);
+
+
+PG_FUNCTION_INFO_V1(get_flattened_wait_graph);
+
+
+/*
+ * get_flattened_wait_graph returns the flattened wait graph. This function
+ * is mostly useful for testing and debugging purposes.
+ */
+Datum
+get_flattened_wait_graph(PG_FUNCTION_ARGS)
```

Comment:
I don't think this UDF adds any value for the user over `dump_global_wait_edges` (also, using a comma-separated string instead of an array would be kind of strange). Could we move it to a test file?